daemon/src/db/time_to_first_position.rs

```diff
@@ -0,0 +1,262 @@
+use anyhow::Result;
+use model::Identity;
+use sqlx::pool::PoolConnection;
+use sqlx::Sqlite;
+use time::OffsetDateTime;
+
+use crate::db::Connection;
+
+impl Connection {
+    /// Record the time at which we hear about a `taker_id` for the
+    /// first time.
+    ///
+    /// If we have already heard about the `taker_id`, we keep the
+    /// original `timestamp`.
+    pub async fn try_insert_first_seen(
+        &self,
+        taker_id: Identity,
+        timestamp: OffsetDateTime,
+    ) -> Result<()> {
+        let mut conn = self.inner.acquire().await?;
+
+        if !have_seen_before(&mut conn, taker_id).await? {
+            insert_first_seen(&mut conn, taker_id, timestamp).await?;
+        }
```

Comment:
Could potentially replace this with SQL?

https://stackoverflow.com/a/19337206

Reply:
nice idea. 👍

Reply:
I think this should work for this instance, but might be a good idea to have another look yourself on how sqlite's conflict resolution algorithm works when using `REPLACE`:

https://sqlite.org/lang_conflict.html

You will definitely need a `UNIQUE` constraint on the `taker_id` then, otherwise `REPLACE` might not work as expected.
Given that we can have `NULL` values in the table please add tests that show that this works as intended!

Reply:
I like this proposal, but `sqlx` doesn't accept the query I've written. Perhaps someone can help me with this.

Reply:
Happy to help out. I'm the master of SQL and killed once a production DB at Siemens using recursive read queries!!!

Reply:
This should do for the `first_seen_timestamp` field.

```
insert or replace into time_to_first_position (taker_id, first_seen_timestamp)
select 'test', 111111
WHERE NOT EXISTS(SELECT id FROM time_to_first_position WHERE taker_id = 'test' and first_seen_timestamp = null);
```

A better solution might actually be to make `first_seen_timestamp` NOT NULL. I think this works because you only insert a taker if you have seen him.

Then you can do this which does not risk overwriting existing values by mistake

```
insert into time_to_first_position (taker_id, first_seen_timestamp)
select 'test', 111111
WHERE NOT EXISTS(SELECT id FROM time_to_first_position WHERE taker_id = 'test' and first_seen_timestamp = null);
```

Reply:
Hold on, this does not work, you might accidentally update the wrong fields 😬

Reply:
So, for inserting the `first_seen_timestamp` you can do this: 

```sql
INSERT OR IGNORE INTO time_to_first_position (taker_id, first_seen_timestamp)
select 'test4', 4
```

It will only insert if it does not exist yet based on the unique key `taker_id`

For updating the first open position you can do: 

```sql
UPDATE time_to_first_position
SET first_position_timestamp = 42
WHERE taker_id = 'test' and first_position_timestamp is null;
```

Reply:
Thanks for giving it a go! I'll try to add it to this PR.

Reply:
Why can't we do the same thing for the `first_position_timestamp`? The approach you propose for the `first_seen_timestamp` seems to work (i.e. tests pass).

Reply:
Afaik, if `taker_id` is already present then `insert or ignore` will ignore it.

Reply:
> Afaik, if `taker_id` is already present then `insert or ignore` will ignore it and won't do anything.

Confirmed!